### ISSUE TITLE
Issues with decorated remotewebdrivers

### ISSUE THREAD

**Issue by adngzz** [SeleniumHQ/selenium#9595]
## 🐛 Bug Report

I am getting an error when trying perform an action(new Actions(decoratedRemoteDriver);) using a decorated remote driver. Also the issue from  case [9416](https://github.com/SeleniumHQ/selenium/issues/9416) about casting back to remotewebdriver from the decorated webdriver is still happening, do i reopen that case?

````
[20][Firefox] Failure: org.openqa.selenium.json.JsonException: java.lang.reflect.InvocationTargetException
Build info: version: '4.0.0-beta-4', revision: '29f46d02dd'
System info: host: 'MTYLP111', ip: '10.1.249.1', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '16'
Driver info: org.openqa.selenium.remote.RemoteWebDriver
Command: [f5415541-00ab-4449-a8f9-939f351b36da, actions {actions=[org.openqa.selenium.interactions.Sequence@4bcade81]}]
Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 89.0, javascriptEnabled: true, moz:accessibilityChecks: false, moz:buildID: 20210527174632, moz:debuggerAddress: localhost:59093, moz:geckodriverVersion: 0.29.1, moz:headless: false, moz:processID: 3480, moz:profile: C:\Users\rgonzalezv\AppData..., moz:shutdownTimeout: 60000, moz:useNonSpecCompliantPointerOrigin: false, moz:webdriverClick: true, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, platformVersion: 10.0, se:cdp: ws://10.1.3.27:4444/session..., se:cdpVersion: 85, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify}
Session ID: f5415541-00ab-4449-a8f9-939f351b36da
[20][Firefox] Stacktrace: org.openqa.selenium.json.SimplePropertyDescriptor.lambda$getPropertyDescriptors$1(SimplePropertyDescriptor.java:107)
org.openqa.selenium.json.JsonOutput.mapObject(JsonOutput.java:392)
org.openqa.selenium.json.JsonOutput.lambda$new$31(JsonOutput.java:207)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.lambda$new$24(JsonOutput.java:182)
java.base/java.util.HashMap.forEach(HashMap.java:1425)
org.openqa.selenium.json.JsonOutput.lambda$new$25(JsonOutput.java:177)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.lambda$new$22(JsonOutput.java:169)
java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
java.base/java.util.LinkedList$LLSpliterator.forEachRemaining(LinkedList.java:1242)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
org.openqa.selenium.json.JsonOutput.lambda$new$23(JsonOutput.java:169)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.lambda$new$24(JsonOutput.java:182)
java.base/java.util.HashMap.forEach(HashMap.java:1425)
org.openqa.selenium.json.JsonOutput.lambda$new$25(JsonOutput.java:177)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.convertUsingMethod(JsonOutput.java:360)
org.openqa.selenium.json.JsonOutput.lambda$new$16(JsonOutput.java:154)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.lambda$new$22(JsonOutput.java:169)
java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
org.openqa.selenium.json.JsonOutput.lambda$new$23(JsonOutput.java:169)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.lambda$new$24(JsonOutput.java:182)
com.google.common.collect.SingletonImmutableBiMap.forEach(SingletonImmutableBiMap.java:65)
org.openqa.selenium.json.JsonOutput.lambda$new$25(JsonOutput.java:177)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:287)
org.openqa.selenium.json.JsonOutput.write(JsonOutput.java:278)
org.openqa.selenium.json.Json.toJson(Json.java:42)
org.openqa.selenium.remote.codec.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:253)
org.openqa.selenium.remote.codec.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:127)
org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:172)
org.openqa.selenium.remote.TracedCommandExecutor.execute(TracedCommandExecutor.java:39)
org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:612)
org.openqa.selenium.remote.RemoteWebDriver.perform(RemoteWebDriver.java:680)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.base/java.lang.reflect.Method.invoke(Method.java:567)
org.openqa.selenium.support.decorators.WebDriverDecorator.call(WebDriverDecorator.java:228)
org.openqa.selenium.support.decorators.DefaultDecorated.call(DefaultDecorated.java:48)
org.openqa.selenium.support.decorators.WebDriverDecorator.lambda$createProxy$0(WebDriverDecorator.java:287)
jdk.proxy2/jdk.proxy2.$Proxy17.perform(Unknown Source)
org.openqa.selenium.interactions.Actions$BuiltAction.perform(Actions.java:617)
org.openqa.selenium.interactions.Actions.perform(Actions.java:573)
com.peektraffic.spinnaker.TestNgTestBase.initDriver(TestNgTestBase.java:591)
com.peektraffic.spinnaker.TestNgTestBase.init(TestNgTestBase.java:164)
com.peektraffic.spinnaker.Help.Test_Help.help_ExplorerIcon(Test_Help.java:139)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.base/java.lang.reflect.Method.invoke(Method.java:567)
org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:133)
org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:598)
org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:173)
org.testng.internal.TestMethodWithDataProviderMethodWorker.call(TestMethodWithDataProviderMethodWorker.java:77)
org.testng.internal.TestMethodWithDataProviderMethodWorker.call(TestMethodWithDataProviderMethodWorker.java:15)
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
java.base/java.lang.Thread.run(Thread.java:831)
````

## To Reproduce

1. Create a new remote webdriver
2. implement a listener
3. decorate the remotewebdriver with EventFiringDecorator
4. Create a new Actions using the decorated driver
5. Perform the action

## Expected behavior
The code should behave the same regardless of if it is done by a remotewebdriver or a decorated remotewebdriver

## Test script or set of commands reproducing this issue
```

		EventListener eventListener = new EventListener();

		
		Long start = System.currentTimeMillis();
		while (true)
		{
			if (System.currentTimeMillis() - start >= MAX_TRY_MS.get())
			{
				fail(threadId.get() + "Couldn't Start driver within " + MAX_TRY_MS.get() + " ms.");
			}
			try
			{
				log(Status.INFO, "Initializing EventFiringWebDriver..");

				remoteDriver = new RemoteWebDriver(new URL(data.get("REMOTE_HUB_LOCATION")), capabilities);

				break;
			}
			catch (Exception e)
			{
				Thread.sleep(5000);
				// log(Status.INFO, "Exception "+ e);
			}

		}
		remoteDriver.get("http://www.google.com");
		Actions Actions = new Actions(remoteDriver);
		Actions.moveToElement(remoteDriver.findElement(By.cssSelector("body")));
		Actions.click();
		Actions.perform();
		// envelop remotedriver into eventfiring webdriver
		log(Status.INFO, "Remotewebdriver session ID : " + remoteDriver.getSessionId());
		log(Status.INFO, "Driver: " + remoteDriver.toString());
		eventDriver.set(new EventFiringDecorator(eventListener).decorate(remoteDriver));
		Actions Actionsdeco = new Actions(eventDriver.get());
		Actionsdeco.moveToElement(eventDriver.get().findElement(By.cssSelector("body")));
		Actionsdeco.click();
                //crashes on the the perform below
		Actionsdeco.perform();
```

## Environment

OS: Windows 10
Browser: Firefox
Browser version: 85
Browser Driver version: GeckoDriver 0.29.1
Language Bindings version: Java Selenium 4 beta 4
Selenium Grid version (if applicable): Selenium 4 beta 4

**Comment by shs96c** [comment]
I wonder what the problem here is? Typically, we should be casting to the interface that we actually need. The `EventFiringWebDriver` is _not_ a `RemoteWebDriver`, but it implements many of the same interfaces.

**Comment by adngzz** [comment]
As far as I know the only purpose of the Eventfiring variants is intercept events for logging purposes and they work as an envelope to whichever type of driver. At least that is the way the old EventFiringWebDriver worked. The new way using decorators seems to only properly support vanilla webdriver. Mostly the problem is the lack of a reliable replacement.

**Comment by pujagani** [comment]
Thank you for providing the details. I was able to reproduce the issue.
The issue has been fixed as part of https://github.com/SeleniumHQ/selenium/commit/4798b008d283602687ebf80ae25048765eae958b and https://github.com/SeleniumHQ/selenium/commit/87fdef6ee602b8dc2a1993e7f4f1a67121dcf482. Thank you @shs96c for the fix!
After trying with the changes from the trunk, I am no longer able to see the issue. It will be available as part of the next release.
